@here/harp-mapview/lib/text/LayoutState.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+    DefaultTextStyle,
+    HorizontalAlignment,
+    TextLayoutStyle,
+    VerticalAlignment
+} from "@here/harp-text-canvas";
+
+/**
+ * Version of layout style stored.
+ */
+export enum LayoutUsed {
+    Undefined = 0,
+    ThemeBased,
+    Alternative
+}
+
+/**
+ * Defines possible text placement relative to anchor.
+ */
+export interface AnchorPlacement {
+    h: HorizontalAlignment;
+    v: VerticalAlignment;
+}
+
+/**
+ * Layout state of the text part of the `TextElement`.
+ *
+ * Used mainly for multi-anchor placement algorithm.
+ * @hidden
+ */
+export class LayoutState {
+    private m_state = LayoutUsed.Undefined;
+    private m_hAlign = DefaultTextStyle.DEFAULT_HORIZONTAL_ALIGNMENT;
+    private m_vAlign = DefaultTextStyle.DEFAULT_VERTICAL_ALIGNMENT;
+
+    /**
+     * Set layout based on theme style defined and optional text placement.
+     *
+     * @param layoutStyle The basic theme label style.
+     * @param placement The optional new anchor placement.
+     */
+    setPlacement(layoutStyle: TextLayoutStyle, placement?: AnchorPlacement) {
+        if (placement !== undefined) {
+            this.m_hAlign = placement.h;
+            this.m_vAlign = placement.v;
+            this.m_state =
+                placement.h === layoutStyle.horizontalAlignment &&
+                placement.v === layoutStyle.verticalAlignment
+                    ? LayoutUsed.ThemeBased
+                    : LayoutUsed.Alternative;
+        } else {
+            this.m_hAlign = layoutStyle.horizontalAlignment;
+            this.m_vAlign = layoutStyle.verticalAlignment;
+            this.m_state = LayoutUsed.ThemeBased;
+        }
+    }
+
+    /**
+     * Acquire current placement setup.
+     *
+     * Function returns alternative or base placement depending on layout state.
+     *
+     * @param layoutStyle The label layout style.
+     * @returns The current anchor placement.
+     */
+    getPlacement(layoutStyle: TextLayoutStyle): AnchorPlacement {
+        if (this.isAlternative()) {
```

Comment:
You're setting in setPlacement the layoutStyle aligment if no alternative placement is given. 
Then why you can't here just get whatever alignment is set in state? It might be an alternative aligment or the one coming from the layout style, you don't care.

Reply:
In fact this things matters, especially when I provide the placement list from theme both with alignment settings. For now it is required even more because of this truly shitty initialization of TextElementState.... We have no guaranty that the layout there is proper, at a very beginning there is no state even... Moreover most of the unit tests (not mine) do not even care about the states! - so you have undefined values there... I need to hold some information about the state stored if it is valid already or anymore, so at least two states are needed -> Initialized / NotInitialized, having 3 states allows to provide more fine tuned logic that Nino has mentioned on the last call... Maybe I will need to reset the state on some scenarios:
- label mostly out of the screen,
- zoom level changed (less or more significantly).
My proposal would be even to make this LayoutState always available so it simplifies a code a lot, but as I saw (from other comments), you rather opt to make it even less commonly set to decrease allocations...

Reply:
In the next commit I will remove all states, let's see how it gonna look & work. Probably everybody should be satisfied.